### PR TITLE
DTS API close out for 2.3

### PR DESCRIPTION
**Pull request by mbolivar-nordic**
DTS changes for 2.3.

This PR is attempting to close out the remaining issues discussed in the last dev-review, namely being able to access data for *all* nodes regardless of their status.

This has knock-on effects in the API which cause various renames.

We're going to defer a few more of them until after code freeze to avoid breaking CI any more.

### CODE REVIEW

File: include/devicetree.h [hunk 1]
```diff
@@ -1396,16 +1401,6 @@
...
 
-/**
- * @brief Does the devicetree have a particular instance number?
- *
- * This is equivalent to DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(inst)).
- * @param inst instance number
- * @return 1 if the devicetree has that numbered instance,
- *         0 otherwise.
- */
-#define DT_HAS_DRV_INST(inst) DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(inst))
```
Comment (galak) [inline include/devicetree.h:1401]: Shouldn't this become `DT_HAS_DRV_INST_STATUS_OKAY(inst)`

**Review by galak — COMMENTED** [review]
Need decide if we want to phase out compat2enabled or tweak its semantics.  I think changing the semantics is ok since the only status values we've really dealt with are (no status, status = "okay", status = "ok", and status = "disabled".  I'm also good if we want to deprecate this and introduce something new.

**Comment by mbolivar-nordic** [comment]
> Need decide if we want to phase out compat2enabled or tweak its semantics. I think changing the semantics is ok since the only status values we've really dealt with are (no status, status = "okay", status = "ok", and status = "disabled". I'm also good if we want to deprecate this and introduce something new.

We agreed in dev-review that whatever we do, we can't break gen_legacy_defines.py. It uses compat2enabled, so we can't change its semantics without also making changes to gen_legacy_defines.py, which seems a bit reckless this late in the release cycle.

I'm adding new functionality specifically to be able to leave the legacy scripts alone.

File: drivers/adc/adc_shell.c [hunk 2]
```diff
@@ -476,13 +453,13 @@ SHELL_STATIC_SUBCMD_SET_CREATE(sub_adc_cmds,
 );
 
 SHELL_STATIC_SUBCMD_SET_CREATE(sub_adc,
-#if DT_HAS_DRV_INST(0)
+#if DT_HAS_NODE_STATUS_OKAY(DT_DRV_INST(0))
```
Comment (galak) [inline drivers/adc/adc_shell.c:456]: I don't think this is correct.  if we have inst0 - ok, inst1 - disabled, inst2-ok, inst3-ok.  The DT_INST_FOREACH_OKAY will get inst0, inst2, inst3.  But dont here we aren't going to match.

File: drivers/clock_control/clock_control_mcux_pcc.c [hunk 3]
```diff
@@ -67,7 +67,12 @@ static const struct clock_control_driver_api mcux_pcc_api = {
 	.get_rate = mcux_pcc_get_rate,
 };
 
-#if DT_HAS_DRV_INST(0)
+#define CHECK_INSTANCE_NUMBER(inst) \
+	BUILD_ASSERT((inst) == 0, "unsupported temp instance");
```
Comment (galak) [inline drivers/clock_control/clock_control_mcux_pcc.c:71]: What's the thinking here?

File: boards/arm/frdm_k64f/pinmux.c [hunk 4]
```diff
@@ -123,7 +123,7 @@ static int frdm_k64f_pinmux_init(struct device *dev)
 #endif
 
 #if DT_HAS_NODE_STATUS_OKAY(DT_NODELABEL(ftm3)) && \
-    DT_NODE_HAS_COMPAT(DT_NODELABEL(ftm3), nxp_kinetis_ftm_pwm)
+    DT_NODE_HAS_COMPAT_STATUS_OKAY(DT_NODELABEL(ftm3), nxp_kinetis_ftm_pwm)
```
Comment (carlescufi) [inline boards/arm/frdm_k64f/pinmux.c:126]: Wouldn't it be better to have the `OKAY` as a parameter to the macro? The name is pretty ugly and to top that it's inconsistent with `DT_INST_FOREACH_OKAY`

File: drivers/clock_control/clock_control_mcux_pcc.c (hunk 3 above)
Comment (mbolivar-nordic) [reply drivers/clock_control/clock_control_mcux_pcc.c:71]: Just trying to move quickly. The driver is single-instance and was hard-coded to instance number 0, so if this blows up, the devicetree is violating the "there's just one instance" assumption and this driver needs to be looked at again.

File: boards/arm/frdm_k64f/pinmux.c (hunk 4 above)
Comment (mbolivar-nordic) [reply boards/arm/frdm_k64f/pinmux.c:126]: Maybe it's better to make DT_INST_FOREACH_OKAY DT_INST_FOREACH_STATUS_OKAY instead. The STATUS_OKAY thing is being used elsewhere.

File: drivers/adc/adc_shell.c (hunk 2 above)
Comment (mbolivar-nordic) [reply drivers/adc/adc_shell.c:456]: changed to:

```
#define ADC_SHELL_COMMAND(inst) \
	SHELL_CMD(ADC_##inst, &sub_adc_cmds, "ADC_" #inst, NULL),

SHELL_STATIC_SUBCMD_SET_CREATE(
	sub_adc,
	DT_INST_FOREACH_OKAY(ADC_SHELL_COMMAND)
	SHELL_SUBCMD_SET_END /* Array terminated. */
);
```

**Comment by galak** [comment]
> > Need decide if we want to phase out compat2enabled or tweak its semantics. I think changing the semantics is ok since the only status values we've really dealt with are (no status, status = "okay", status = "ok", and status = "disabled". I'm also good if we want to deprecate this and introduce something new.
> 
> We agreed in dev-review that whatever we do, we can't break gen_legacy_defines.py. It uses compat2enabled, so we can't change its semantics without also making changes to gen_legacy_defines.py, which seems a bit reckless this late in the release cycle.
> 
> I'm adding new functionality specifically to be able to leave the legacy scripts alone.

We should probably add a comment in compat2enabled and node.enabled about these only existing for gen_legacy_defines.py.  Do we want to call this compat2avail to match the linux API?

File: scripts/dts/gen_defines.py [hunk 5]
```diff
@@ -359,6 +353,15 @@ def write_compatibles(node):
             f"{node.z_path_id}_COMPAT_MATCHES_{str2ident(compat)}", 1)
 
 
+def write_status(node):
+    status = node.props.get('status')
```
Comment (galak) [inline scripts/dts/gen_defines.py:355]: Use node.status now.

File: scripts/dts/gen_defines.py [hunk 6]
```diff
@@ -359,6 +353,15 @@ def write_compatibles(node):
...
+
+    if status is None:
+        return
```
Comment (galak) [inline scripts/dts/gen_defines.py:358]: We can either remove this check or make it a warning as `node.status` should never be none.

**Comment by mbolivar-nordic** [comment]
> Do we want to call this compat2avail to match the linux API?

Do you also want to change all the STATUS_OKAY macros to say AVAIL? Would be best to keep those consistent in my opinion.

**Comment by pabigot** [comment]
> > Do we want to call this compat2avail to match the linux API?
> 
> Do you also want to change all the STATUS_OKAY macros to say AVAIL? Would be best to keep those consistent in my opinion.

I think I support that.  Or `_AVAILABLE`.

From what I can see Linux really uses this to mean `status = "okay"`, so we should keep that interpretation.  To handle #19448 we'll need some other technical term for things for which devices should be generated, but that should not be started in the boot-time driver initialization.  I don't think there's a term for that in Linux because everything in the devicetree blob is "available" (in the general sense).

**Comment by mbolivar-nordic** [comment]
I'm going to leave this as STATUS_OKAY to avoid yet more churn unless someone objects.

File: scripts/dts/edtlib.py [hunk 7]
```diff
@@ -142,7 +142,8 @@ class EDT:
       The bindings directory paths passed to __init__()
     """
     def __init__(self, dts, bindings_dirs, warn_file=None,
-                 warn_reg_unit_address_mismatch=True):
+                 warn_reg_unit_address_mismatch=True,
+                 default_prop_types=None):
```
Comment (galak) [inline scripts/dts/edtlib.py:156]: This means ever user (kconfiglib, sanitycheck, etc) has to define the same set of default props?  That doesn't make sense to me.

**Review by galak — COMMENTED** [review]
Would be good to update doc/guides/dts/macros.bnf in this PR so we don't forget.

**Comment by mbolivar-nordic** [comment]
> Would be good to update doc/guides/dts/macros.bnf in this PR so we don't forget.

Pushed various docs updates.

File: scripts/dts/edtlib.py [hunk 8]
```diff
@@ -160,12 +161,17 @@ def __init__(self, dts, bindings_dirs, warn_file=None,
           If True, a warning is printed if a node has a 'reg' property where
           the address of the first entry does not match the unit address of the
           node
+
+        default_prop_types (default: False):
+          If True, default property types will be used when a node has no
```
Comment (galak) [inline scripts/dts/edtlib.py:176]: Can we default this to True and update the old script to call it with False.

Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py:176]: > Can we default this to True and update the old script to call it with False.

done

File: scripts/dts/gen_defines.py [hunk 9]
```diff
@@ -36,7 +36,8 @@ def main():
         edt = edtlib.EDT(args.dts, args.bindings_dirs,
                          # Suppress this warning if it's suppressed in dtc
                          warn_reg_unit_address_mismatch=
-                             "-Wno-simple_bus_reg" not in args.dtc_flags)
+                             "-Wno-simple_bus_reg" not in args.dtc_flags,
+                         default_prop_types=False)
```
Comment (galak) [inline scripts/dts/gen_defines.py]: Don't you want to remove default_prop_types here.

File: scripts/dts/edtlib.py [hunk 10]
```diff
@@ -160,12 +161,17 @@ def __init__(self, dts, bindings_dirs, warn_file=None,
           If True, a warning is printed if a node has a 'reg' property where
           the address of the first entry does not match the unit address of the
           node
+
+        default_prop_types (default: False):
```
Comment (galak) [inline scripts/dts/edtlib.py:175]: Need to fix comment to match the code `(default: False) -> (default: True)`

Comment (mbolivar-nordic) [reply scripts/dts/edtlib.py]: Thanks for catching that

File: scripts/dts/gen_defines.py (hunk 9 above)
Comment (mbolivar-nordic) [reply scripts/dts/gen_defines.py]: It becomes True later on in the series (`devicetree: allow access to all nodes`). I'm just trying to preserve bisectability of behavior.

**Comment by mbolivar-nordic** [comment]
Will fix the shippable issues in the morning.

**Review by galak — APPROVED** [review]

**Comment by galak** [comment]
We need to look at DT_ANY_INST_ON_BUS_STATUS_OKAY  -- drivers/sensor/iis2dlpc fails to build on frdm_k64f.  Since DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) is true, but INST(0) is not enabled.

**Comment by galak** [comment]
We need fix 25059 and I've got a fix for the nxp_kinetis_temp

**Comment by galak** [comment]
> We need to look at DT_ANY_INST_ON_BUS_STATUS_OKAY -- drivers/sensor/iis2dlpc fails to build on frdm_k64f. Since DT_ANY_INST_ON_BUS_STATUS_OKAY(spi) is true, but INST(0) is not enabled.

So I think for singleton drivers we need something that gives back the list of OKAY INST numbers and than we can use the first item in that list for the singletons

File: drivers/can/can_stm32.c [hunk 11]
```diff
@@ -20,10 +20,8 @@ LOG_MODULE_DECLARE(can_driver, CONFIG_CAN_LOG_LEVEL);
 
 #define CAN_INIT_TIMEOUT  (10 * sys_clock_hw_cycles_per_sec() / MSEC_PER_SEC)
 
-#if DT_HAS_NODE_STATUS_OKAY(DT_NODELABEL(can1)) && \
-	DT_NODE_HAS_COMPAT(DT_NODELABEL(can1), st_stm32_can) && \
-	DT_HAS_NODE_STATUS_OKAY(DT_NODELABEL(can2)) && \
-	DT_NODE_HAS_COMPAT(DT_NODELABEL(can2), st_stm32_can)
+#if DT_NODE_HAS_COMPAT_STATUS_OKAY(DT_NODELABEL(can1), st_stm32_can) && \
```
Comment (erwango) [inline drivers/can/can_stm32.c:23]: While it should work, it doesn't:
`sanitycheck -b -T  samples/drivers/can/ -p olimex_stm32_p405`
```
/local/mcu/zephyrproject/zephyr/drivers/can/can_stm32.c:25:2: error: #error Simultaneous use of CAN_1 and CAN_2 not supported yet
   25 | #error Simultaneous use of CAN_1 and CAN_2 not supported yet
      |  ^~~~~

```

**Review by erwango — DISMISSED** [review]

File: drivers/can/can_stm32.c (hunk 11 above)
Comment (mbolivar-nordic) [reply drivers/can/can_stm32.c]: @erwango fixed now, please take another look.

**Review by pabigot — APPROVED** [review]
Haven't looked closely at the documentation yet but figure that can be tuned if necessary.

**Comment by mbolivar-nordic** [comment]
> Haven't looked closely at the documentation yet but figure that can be tuned if necessary.

Yep, I promise to go over the documentation with a fine tooth comb before the release and correct any errors I find during the stabilization/documentation period. Thanks.

File: doc/guides/dts/howtos.rst [hunk 12]
```diff
@@ -104,8 +104,7 @@ a node identifier and pass it to ``DT_LABEL`` to get the right string to pass
 to ``device_get_binding()``.
 
 If you're having trouble, see :ref:`dt-trouble`. The first thing to check is
-that the node is enabled (``status = "okay"``) and has a matching binding, like
-this:
+that the node is has ``status = "okay"`` and has a matching binding, like this:
```
Comment (ioannisg) [inline doc/guides/dts/howtos.rst:107]: remove "is"

File: doc/guides/dts/howtos.rst [hunk 13]
```diff
@@ -114,11 +113,11 @@ this:
    #if DT_HAS_NODE_STATUS_OKAY(MY_SERIAL)
    struct device *uart_dev = device_get_binding(DT_LABEL(MY_SERIAL));
    #else
-   #error "Node is disabled or has no matching binding"
+   #error "Node is disabled, has no matching binding, or initialization failed"
    #endif
 
-If you see the ``#error`` output, something is wrong with either your
-devicetree or bindings.
+If you see the ``#error`` output, either something is wrong with either your
```
Comment (ioannisg) [inline doc/guides/dts/howtos.rst:119]: remove second either in the line

**Review by ioannisg — APPROVED** [review]
Looks ok to me
Pls, address some minor doc-related comments.

File: doc/guides/dts/howtos.rst (hunk 13 above)
Comment (pabigot) [reply doc/guides/dts/howtos.rst:119]: No, this is correct.  There are two optional subordinate clauses, one of which requires its own `either`.

Comment (pabigot) [reply doc/guides/dts/howtos.rst:119]: Or rather, it's correct with two eithers, just the first, just the second, or neither.  IMO.  English is pretty flexible.

**Review by erwango — APPROVED** [review]

**Pull request merged**